code/photometry.py

```diff
@@ -0,0 +1,168 @@
+                         NullPhotoCal, ConstantSky, Image)
+
+except ImportError:
+    print("tractor is missing")
+    pass
+
+
+def calc_background(*, bkgsubimage):
+    """Measure sky noise and mean level.
+
+    Parameters:
+    -----------
+    bkgsubimage: <need type>
+        Image data from which the sky background will be calculated.
+
+    Returns:
+    --------
+    skymean: float
+        Mean of sigma-clipped background
+    skynoise: float
+        Standard deviation of sigma-clipped background
+    """
+    # suppress warnings about nans in the calculation
+    with warnings.catch_warnings():
+        warnings.simplefilter("ignore", category=RuntimeWarning)
+        skymean, skymedian, skynoise = sigma_clipped_stats(
+            bkgsubimage, sigma=3.0)
+
+    return skymean, skynoise
+
+
+def run_tractor(*, subimage, prf, objsrc, skymean, skynoise):
+    """Make the tractor image and perform forced photometry.
+
+    Parameters:
+    -----------
+    subimage: <need type>
+        Science image cutout.
+    prf: np.ndarray
+        Point spread function for the band/channel.
+    objsrc: List[tractor.ducks.Source]
+        List of tractor Source objects for the target and nearby sources.
+    skymean: float
+        Mean of sigma-clipped background
+    skynoise: float
+        Standard deviation of sigma-clipped background
+
+    Returns:
+    -------
+    flux_var: float, double, or None
+        Flux variance result from the tractor optimization.
+        None if tractor optimization succeeded but it didn't report a variance.
+
+    Raises:
+    -------
+    TractorError: If the tractor optimization fails.
+    """
+    # make the tractor image
+    tim = Image(
+        data=subimage,
+        invvar=np.ones_like(subimage) / skynoise**2,
+        psf=PixelizedPSF(prf),
+        wcs=NullWCS(),
+        photocal=NullPhotoCal(),
+        sky=ConstantSky(skymean),
+    )
+
+    # make tractor object combining tractor image and source list
+    tractor = Tractor([tim], objsrc)  # [src]
+
+    # freeze the parameters we don't want tractor fitting
+    tractor.freezeParam("images")  # now fits 2 positions and flux
+    # tractor.freezeAllRecursive()#only fit for flux
+    # tractor.thawPathsTo('brightness')
+
+    # run the tractor optimization (do forced photometry)
+    # Take several linearized least squares steps
+    try:
+        tr = 0
+        with suppress_stdout():
+            with warnings.catch_warnings():
+                warnings.filterwarnings("ignore", ".*divide by zero.*")
+                # warnings.simplefilter('ignore')
+                for tr in range(20):
+                    dlnp, X, alpha, flux_var = tractor.optimize(variance=True)
+                    # print('dlnp',dlnp)
+                    if dlnp < 1e-3:
+                        break
+
+    # catch exceptions and bad fits
+    except Exception as e:
+        raise TractorError("Tractor failed to converge") from e
```

Comment:
again, how meaningful the tractor traceback, do we need to have this custom one here and the try/except?

Reply:
The custom errors were just a simple way to isolate these try-except blocks so that I could refactor and test the rest of the code more easily. They are mostly about making the development easier, and can certainly be taken out when we fix the try-except logic (but fixing that logic was outside the scope of this PR). I explained more of my thought process in my comment above about the `CutoutError`. That said, I'm certainly interested to know if you would prefer a different strategy here (short of fixing the try-except logic itself)? 

Bigger picture:
I suspect that tractor really is failing sometimes (unlike the cutouts), and that it's fine to just skip that source when it does fail. So I think it probably makes sense to have a try-except block here so that it doesn't crash the program entirely. IMO, the custom `TractorError` is useful because it lets the main function catch a specific error without being tied up in the details of what the error is -- so everything else can move on while we try to sort out those details. Once we do sort it out we can replace the custom error with whatever tractor is actually throwing.

Reply:
> Bigger picture:

OK, I like your bigger picture explanation and don't mind to have the custom one if the user cannot really do anything to fix those sources. But I still do mind the generic exception, presumably tractor raises specific types, and only those (maybe even matching for the error messages rather than letting all e.g. ValueError through), should be captured. 

But I don't think at this point it's worth spending time to run tractor again waiting for these errors, so it's OK to leave this as is and keep this rule of thumb of capturing specific, and understood things only, in mind for the new use cases.